### ISSUE TITLE
": Undefined location" when running a manual Mutation (next branch) + Random 127.0.0.1 connections

### ISSUE THREAD

**Issue by eliezedeck** [zino-app/graphql-flutter#186]
**Describe the bug**
When running a mutation manually, I get 2 errors at different code levels:
- `SocketException: OS Error: Connection refused, errno = 111, address = localhost, port = 33860` with random port (location is: `flutter_graphql/src/link/link_http.dart:85`) ... this is directly related to the mutation that I'm trying to run
- `: Undefined location` after getting the awaited result of the mutation. Note that my backend is on localhost.

I was trying to get to the code that caused connections to random ports, but I couldn't use the IDE to pin-point the source of the problem. Makes me wonder, why is it doing this random connection and not connecting to where it's supposed to.

**To Reproduce**
1. Start any project that would use `flutter_graphql` (same exact behavior when using `graphql_flutter`'s `next` branch or 1.0.0 beta)
2. Use an `HttpLink` that is pointing to a fixed `http://localhost:8080/v1/graphql`
3. At this point, any manual mutations I tried to run would cause the same exact error
4. Even absolutely invalid document will also return the same error
5. Set a break-point deeper and deeper, and eventually, you'll meet `SocketException`

**Expected behavior**
A connection is made to my GraphQL backend. Unfortunately, my backend stays still, not receiving any connection at all.

**Screenshots**
Here are some screens:

![image](https://user-images.githubusercontent.com/7553916/54153225-1ca29400-4450-11e9-8305-a255df3c52dc.png)

![image](https://user-images.githubusercontent.com/7553916/54153196-07c60080-4450-11e9-95ab-0ad5a81a1fbd.png)

**Desktop (please complete the following information):**
 - OS: Linux Mint (based on Ubuntu)
 - Version: 19.1

**Smartphone (please complete the following information):**
 - Device: Emulator
 - OS: Android
 - Version: any

**Additional context**
I was looking at the 0.9.5 version, but there was no way to run manual Mutations.

**Comment by chirdeeptomar** [comment]
Same Error: 
graphql_flutter: ^1.0.0-beta

```
   final HttpLink link = HttpLink(
      uri: 'http://127.0.0.1:8000/graphql',
    );

    GraphQLClient graphQl = GraphQLClient(
      cache: InMemoryCache(),
      link: link,
    );
    String getAllCustomers = """
          query getCustomers {
            allCustomers {
              id
              name
            }
          }
        """
        .replaceAll('\n', ' ');

    QueryOptions q = QueryOptions(document: getAllCustomers);
    var result = await graphQl.query(q);
    print(result.errors);
```

**Comment by eliezedeck** [comment]
I mean, is this project still alive? At least, a quick reply would be appreciated. Thanks.

**Comment by chirdeeptomar** [comment]
Don't think it is....https://github.com/snowballdigital/flutter-graphql but still the same problem!

**Comment by yunyu** [comment]
This only occurs when there is an error returned in the mutation, at least with snowballdigital's fork. This catch statement (https://github.com/snowballdigital/flutter-graphql/blob/master/lib/src/core/query_manager.dart#L143) breaks, as the error object is a builtin dart error that has no message getter.

**Comment by eliezedeck** [comment]
> This only occurs when there is an error returned in the mutation

But that doesn't explain why no connection is made to my GraphQL server, at all.

> If there's no error, the mutation succeeds and the data is returned correctly.

Logically, the framework will never know if there was error unless it connects to the GraphQL first.

The main problem here is that the Flutter app using the library never connects to the actual GraphQL server; instead, it tries to connect to random ports. The question of why it does that is still not answered.

**Comment by yunyu** [comment]
@eliezedeck I was able to connect to my server fine - I'm hosting it remotely, as I was unable to get localhost binding working. I'm fairly sure that the issue you reported happens whenever *any* error occurs, and we were experiencing separate issues. Can you connect to your localhost endpoint using just the Dart HTTP client?

**Comment by micimize** [comment]
@yunyu to be clear, you were unable to connect to `localhost:port`, even from simulators? 
This is probably mostly some kind of [firewall or networking issue](https://stackoverflow.com/questions/53342939/http-request-not-processing-while-emulating-or-running-on-real-device-mode-but-w)

The ["random ports" ](https://stackoverflow.com/questions/47372568/how-to-point-to-localhost8000-with-the-dart-http-package-in-flutter?rq=1) can happen as a result of network issues. I'm investigating similar connection issues with `iOS`. A temporary workaround might be to use [`ngrok`](https://ngrok.com/)

To keep track of this, here are the environs we're trying to resolve this for:
- [ ] android simulator - setting host to `10.0.2.2`is the main candidate solution 
- [ ] android device - `adb reverse` should work
- [ ] iOS device
- [ ] iOS simulator - works for me with `localhost` and `127.0.0.1	localhost local` in my `/etc/hosts`, idk if there are other configurations I needed to make

**Comment by yunyu** [comment]
@micimize We didn't bother looking up the 10.0.2.2 and localhost difference for Android and iOS, and somehow ended up using the same one for testing. That said, I am running into this bug (or at least a variant of it) even when there are no networking issues - errors with a status code of 400 but returning a valid GraphQL response are "eaten" with 400 Bad Request (such as Apollo syntax/data format errors).

**Comment by micimize** [comment]
@yunyu exception swallowing is a different issue - at least I'm going to treat it that way, it's kind of a pervasive issue atm but we've got merged changes that should make it better.

Anyways, please provide more details in #177

**Comment by yunyu** [comment]
Just kidding, this is an issue with the current next branch of this library, but not with https://github.com/snowballdigital/flutter-graphql. I just did a quick conversion between the two, and flutter_graphql worked without a SocketError while rev `89fc23c45becd9a8fb90fbede734282941ec531c` of graphql_flutter did not.

**Comment by micimize** [comment]
So, I think @yunyu's PR fixed one root cause, and I made #212 for tracking this generally. Closing this, if y'all encounter it again please comment on #212 or make a new issue with more specific details (operation string, port server is running on, environment you're trying to connect from.
